src/app/student/map/_components/MapComponent.tsx

```diff
@@ -0,0 +1,208 @@
+"use client"
+
+import { BoothPopup } from "@/app/student/map/_components/BoothPopup"
+import {
+  BoothID,
+  geoJsonBoothDataByLocation,
+  geoJsonBuildingData
+} from "@/app/student/map/lib/booths"
+import { Location } from "@/app/student/map/lib/locations"
+import "maplibre-gl/dist/maplibre-gl.css"
+import { useEffect, useMemo, useRef, useState } from "react"
+import {
+  Layer,
+  Map as MapboxMap,
+  MapLayerMouseEvent,
+  MapRef,
+  Source
+} from "react-map-gl/maplibre"
+import { BoothMap, GeoJsonBooth } from "../lib/booths"
+import { BoothMarker } from "./BoothMarker"
+import {
+  backgroundLayerStyle,
+  boothLayerStyle,
+  buildingLayerStyle
+} from "../lib/config"
+
+export function MapComponent({
+  boothsById,
+  location,
+  activeBoothId,
+  setActiveBoothId,
+  hoveredBoothId,
+  setHoveredBoothId,
+  initialView,
+  filteredBoothIds
+}: {
+  boothsById: BoothMap
+  location: Location
+  activeBoothId: BoothID | null
+  hoveredBoothId: BoothID | null
+  setActiveBoothId: (id: BoothID | null) => void
+  setHoveredBoothId: (id: BoothID | null) => void
+  initialView: { longitude: number; latitude: number; zoom: number }
+  filteredBoothIds: BoothID[]
+}) {
+  const mapRef = useRef<MapRef>(null)
+
+  const [markerScale, setMarkerScale] = useState(1)
+
+  function flyToLocation(location: Location) {
+    const { longitude, latitude, zoom } = location.center
+    mapRef.current?.flyTo({
+      center: [longitude, latitude],
+      zoom: zoom
+    })
+  }
+
+  // Fly to location center on change
+  useEffect(() => flyToLocation(location), [location])
+
+  // Fly to selected booth on change
+  useEffect(() => {
+    if (activeBoothId == null) return
+    const booth = boothsById.get(activeBoothId)
+    if (!booth) return
+
+    mapRef.current?.flyTo({
+      center: booth.center as [number, number],
+      zoom: 18.5,
+      speed: 0.8
+    })
+  }, [activeBoothId, boothsById])
+
+  // Keep mapbox feature state in sync with component state
+  // to allow for styling of the features
+  function useFeatureState(
+    boothIds: BoothID[],
+    stateKey: "active" | "hover" | "filtered"
+  ) {
+    useEffect(() => {
+      const map = mapRef.current
+      if (map == null || boothIds.length === 0) return
+
+      for (const boothId of boothIds) {
+        map.setFeatureState(
+          { source: "booths", id: boothId },
+          { [stateKey]: true }
+        )
+      }
+
+      return () => {
+        for (const boothId of boothIds) {
+          map.setFeatureState(
+            { source: "booths", id: boothId },
+            { [stateKey]: false }
+          )
+        }
+      }
+    }, [boothIds, stateKey])
+  }
```

Comment:
Sorry did you mean extracting the entire useFeatureState function, or something else? Also just curious, how could it cause rerenders, isn't it just a function definition

Reply:
I think this video explains it quite well: https://www.youtube.com/watch?v=2sAdzy90GtE

But in essence, its problematic to have components inside other components (in this case a hook inside a component but it is the same thing) because it will cause the hook to re-render on every parent rerender